time-series-utils/preprocessing/script.whizzml

```diff
@@ -0,0 +1,145 @@
+;; information and its name
+
+(define (rows-to-dataset rows fields ds-name)
+  (let (fields (sort-by-key "column_number" (values fields))         
+        headers (map (lambda (x) (x "name")) fields)
+        data (join "\n" (map (lambda (r) (join "," r)) (cons headers rows)))
+        source (create-source {"data" data})
+        dataset (create-and-wait-dataset source {"name" ds-name}))
+    (delete (wait source))
+    dataset))
+
+(define (fill-datetime ds-id key-field num-key-field)
+  (create-dataset {"origin_dataset" ds-id
+                   "all_but" [key-field]
+                   "new_fields"
+                     [{"names" [key-field]
+                       "field"
+                       (flatline "(if (missing? {{key-field}})"
+                                 "(str (epoch-year (f {{num-key-field}})) \"-\" "
+                                 " (epoch-month (f {{num-key-field}})) \"-\" "
+                                 " (epoch-day (f {{num-key-field}})))"
+                                 "(f {{key-field}}))")
+                       "optype" "datetime"}]}))
+                                 
+;; Extracts the rows in a dataset that contains an epoch datetime key field
+;; fills its gaps and removes duplicates in the key field according to the 
+;; selected lapse
+;; Inputs:
+;;   dataset-id - The dataset ID
+;;   key-field - Name of the datetime or numeric field used as key
+;;   lapse - Number of milliseconds that should separate each row   
+;;   dt-field-format - Format to parse the datetime key field (if needed). If empty, "YYYY-MM-dd" will be used.
+(define (ts-normalize dataset-id key-field lapse dt-field-format)
+  (let (fields (resource-fields dataset-id) 
+        name ((fetch dataset-id) "name")
+        sorted-name (str "non-decreasing " name)
+        key-field-info (find-field fields key-field)
+        dt-field-format (if (= "") "YYYY-MM-dd" dt-field-format)
+        [key-field num-key-field ds-id] (create-sorted-asc dataset-id
+                                                           key-field-info
+                                                           dt-field-format
+                                                           sorted-name)
+        fields (resource-fields (wait ds-id))
+        key-col ((find-field fields key-field) "column_number")
+        num-key-col ((find-field fields num-key-field) "column_number")
+        ds-name (str "ts-normalized " name)
+        rows (fetch-dataset-rows ds-id)
+        rows (first rows num-key-col)
+        rows (fill-gaps rows key-col num-key-col lapse)
+        norm-ds (rows-to-dataset rows fields ds-name))
+        (if (= key-col num-key-col)
+            norm-ds
+            (fill-datetime norm-ds key-field num-key-field))))
+       
+(define normalized-dataset (wait (ts-normalize dataset-id key-field lapse dt-field-format)))
+(for (id (created-resources)) (if (!= id normalized-dataset) (delete id)))
+normalized-dataset
```

Comment:
one could set a flag in the script definition marking all created resources as temp (outputs are automatically excluded)